Analysis completed but looks incomplete

Dev 15 RC with Anaconda
I've started/restarted VS a few times during analysis.
Now it appears to be done, but it's not recognizing simple things such as `unittest` module (I've reloaded solution, then restarted VS but same result - not recognized)

![image](https://cloud.githubusercontent.com/assets/1696845/19864393/4e843354-9f55-11e6-8af3-89047ade7227.png)

Full diagnostics on our bug share `Diagnostic Info 2016-10-31T102711.txt`.

These errors appear several times:
```
Traceback (most recent call last):
  File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\Extensions\Microsoft\Python\Core\PythonScraper.py", line 390, in generate_member
    return 'method', generate_builtin_function(obj, True)
  File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\Extensions\Microsoft\Python\Core\PythonScraper.py", line 275, in generate_builtin_function
    function_table['overloads'] = BuiltinScraper.get_overloads(function, is_method)
  File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\Extensions\Microsoft\Python\Core\BuiltinScraper.py", line 519, in get_overloads
    extra_args,
  File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\Extensions\Microsoft\Python\Core\BuiltinScraper.py", line 476, in get_overloads_from_doc_string
    res = parse_doc_str(doc_str, mod, decl_mod, func_name, extra_args, obj_class)
  File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\Extensions\Microsoft\Python\Core\BuiltinScraper.py", line 267, in parse_doc_str
    ret_types = get_ret_type(ret_type, obj_class, mod)
  File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\Extensions\Microsoft\Python\Core\BuiltinScraper.py", line 186, in get_ret_type
    return [type_name_to_typeref(ret_type, mod, RETURN_TYPE_OVERRIDES)]
  File "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\Extensions\Microsoft\Python\Core\BuiltinScraper.py", line 144, in type_name_to_typeref
    for mod_name, mod in sys.modules.items():
RuntimeError: dictionary changed size during iteration
```

```
2016-10-31T10:22:33: [ERROR] Analysis failed
System.ArgumentNullException: Value cannot be null.
Parameter name: key
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GetTypeName(String moduleName, String className)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GenerateTypeName(AnalysisValue baseClass, Boolean isRef)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GenerateTypeName(AnalysisValue baseClass, Boolean isRef)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.<>c__DisplayClass35_0.<GenerateTypeName>b__0(AnalysisValue ns)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.<DistinctIterator>d__63`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GenerateTypeName(IAnalysisSet name, Boolean isRef)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GenerateParameter(Parameter param, IAnalysisSet typeInfo)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GenerateArgInfo(FunctionInfo fi, IAnalysisSet[] parameters)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GenerateOverloads(FunctionInfo fi)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GenerateFunction(FunctionInfo fi)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GetMemberValueInternal(AnalysisValue type, ModuleInfo declModule, Boolean isRef)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GetMemberValue(IAnalysisSet types, ModuleInfo declModule, Boolean isRef)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GenerateMember(VariableDef variableDef, ModuleInfo declModule, Boolean isRef)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.GenerateMembers(ModuleInfo moduleInfo, IEnumerable`1 children)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.SerializeModule(ModuleInfo moduleInfo)
   at Microsoft.PythonTools.Analysis.SaveAnalysis.Save(PythonAnalyzer state, String outDir)
   at Microsoft.PythonTools.Analysis.PyLibAnalyzer.Analyze()
   at Microsoft.PythonTools.Analysis.PyLibAnalyzer.<RunWorker>d__33.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.PythonTools.Analysis.PyLibAnalyzer.<Run>d__32.MoveNext()
```

--- a/Python/Product/Analysis/SaveAnalysis.cs
+++ b/Python/Product/Analysis/SaveAnalysis.cs
@@ -60,7 +60,6 @@ namespace Microsoft.PythonTools.Analysis {
                         WriteModule(outDir, name, info, moduleInfo.Scope.AllVariables.Keys());
                     } catch (Exception ex) {
                         AddError(ex.ToString());
-                        _errorStack.Clear();
                     }
                 }
             }

--- a/Python/Product/Analysis/SaveAnalysis.cs
+++ b/Python/Product/Analysis/SaveAnalysis.cs
@@ -28,12 +28,19 @@ using Microsoft.PythonTools.Parsing.Ast;
 
 namespace Microsoft.PythonTools.Analysis {
     class SaveAnalysis {
-        private List<string> _errors = new List<string>();
-        private Dictionary<AnalysisValue, string> _classNames = new Dictionary<AnalysisValue, string>();
-        private List<AnalysisValue> _path = new List<AnalysisValue>();
-        private Dictionary<string, Dictionary<string, object>> _typeNames = new Dictionary<string, Dictionary<string, object>>();
-        private Dictionary<string, string> _MemoizedStrings = new Dictionary<string, string>();
-        private Dictionary<string, object[]> _moduleNames = new Dictionary<string, object[]>();
+        // This stack contains all module, class and member names while saving
+        // analysis. It should only be popped on successful generation (that is,
+        // do not use a 'finally' block) so the top-level error reporting will
+        // include the full C# call stack as well as the name of the member when
+        // the error occurred.
+        private readonly Stack<string> _errorStack = new Stack<string>();
+        private readonly List<string> _errors = new List<string>();
+
+        private readonly Dictionary<AnalysisValue, string> _classNames = new Dictionary<AnalysisValue, string>();
+        private readonly List<AnalysisValue> _path = new List<AnalysisValue>();
+        private readonly Dictionary<string, Dictionary<string, object>> _typeNames = new Dictionary<string, Dictionary<string, object>>();
+        private readonly Dictionary<string, string> _MemoizedStrings = new Dictionary<string, string>();
+        private readonly Dictionary<string, object[]> _moduleNames = new Dictionary<string, object[]>();
         private static readonly List<object> _EmptyMro = new List<object>();
         private PythonAnalyzer _curAnalyzer;
         private ModuleInfo _curModule;
@@ -46,14 +53,23 @@ namespace Microsoft.PythonTools.Analysis {
                 ModuleInfo moduleInfo;
                 if ((moduleInfo = modKeyValue.Value.Module as ModuleInfo) != null) {
                     _curModule = moduleInfo;
-                    var info = SerializeModule(moduleInfo);
-                    WriteModule(outDir, name, info, moduleInfo.Scope.AllVariables.Keys());
+                    _errorStack.Clear();
+                    _errorStack.Push(modKeyValue.Key);
+                    try {
+                        var info = SerializeModule(moduleInfo);
+                        WriteModule(outDir, name, info, moduleInfo.Scope.AllVariables.Keys());
+                    } catch (Exception ex) {
+                        AddError(ex.ToString());
+                        _errorStack.Clear();
+                    }
                 }
             }
+        }
 
-            foreach (var error in _errors) {
-                Console.WriteLine(error);
-            }
+        public IEnumerable<string> Errors => _errors;
+
+        private void AddError(string message) {
+            _errors.Add("{0}: {1}".FormatInvariant(string.Join(".", _errorStack.Reverse()), message));
         }
 
         private void WriteModule(string outDir, string name, Dictionary<string, object> info, IEnumerable<string> globals) {
@@ -138,7 +154,9 @@ namespace Microsoft.PythonTools.Analysis {
                     // Never got a value, so leave it out.
                     continue;
                 }
+                _errorStack.Push(keyValue.Key);
                 res[keyValue.Key] = GenerateMember(keyValue.Value, moduleInfo);
+                _errorStack.Pop();
             }
             foreach (var child in children.OfType<string>()) {
                 object modRef = null;
@@ -319,7 +337,7 @@ namespace Microsoft.PythonTools.Analysis {
                 var type = types.First();
                 var res = GetMemberValueInternal(type, declModule, isRef);
                 if (res == null) {
-                    _errors.Add(String.Format("Cannot save single member: {0}", types.First()));
+                    AddError(String.Format("Cannot save single member: {0} ({1})", type, type.GetType()));
                 }
                 return res;
             } else if (types.Count == 0) {
@@ -508,7 +526,8 @@ namespace Microsoft.PythonTools.Analysis {
         }
 
         private Dictionary<string, object> GenerateClass(ClassInfo ci, ModuleInfo declModule) {
-            return new Dictionary<string, object>() {
+            _errorStack.Push(ci.Name);
+            var res = new Dictionary<string, object>() {
                 { "mro", GetClassMro(ci) },
                 { "bases", GetClassBases(ci) },
                 { "members", GetClassMembers(ci, declModule) },
@@ -516,6 +535,8 @@ namespace Microsoft.PythonTools.Analysis {
                 { "builtin", false },
                 { "location", GenerateLocation(ci.Locations) }
             };
+            _errorStack.Pop();
+            return res;
         }
 
         private object GetTypeRef(string moduleName, string className) {
@@ -523,6 +544,11 @@ namespace Microsoft.PythonTools.Analysis {
         }
 
         private object GetTypeName(string moduleName, string className) {
+            if (moduleName == null || className == null) {
+                AddError("GetTypeName({0}, {1})".FormatInvariant(moduleName ?? "(null)", className ?? "(null)"));
+                return null;
+            }
+            
             // memoize types names for a more efficient on disk representation.
             object typeName;
             Dictionary<string, object> typeNames;
@@ -588,7 +614,7 @@ namespace Microsoft.PythonTools.Analysis {
                 return GenerateTypeName(name.First(), isRef);
             }
 
-            return name.Select(ns => GenerateTypeName(ns, isRef)).Distinct().ToList<object>();
+            return name.Select(ns => GenerateTypeName(ns, isRef)).Where(n => n != null).Distinct().ToList<object>();
         }
 
         private object GenerateTypeName(AnalysisValue baseClass, bool isRef) {
@@ -622,6 +648,10 @@ namespace Microsoft.PythonTools.Analysis {
 
         private object GenerateTypeName(IPythonType type) {
             if (type != null) {
+                if (string.IsNullOrEmpty(type.Name)) {
+                    AddError("{0} has no name".FormatInvariant(type.TypeId));
+                    return null;
+                }
                 return GetTypeName(type.DeclaringModule.Name, type.Name);
             }
             return null;

--- a/Python/Product/Analyzer/BuiltinScraper.py
+++ b/Python/Product/Analyzer/BuiltinScraper.py
@@ -141,7 +141,7 @@ def type_name_to_typeref(name, mod, type_overrides = TYPE_OVERRIDES):
             arg_type = PythonScraper.type_to_typeref(list)
         else:
             # see if we can find it in any module we've imported...
-            for mod_name, mod in sys.modules.items():
+            for mod_name, mod in list(sys.modules.items()):
                 if mod is not None and name in mod.__dict__ and isinstance(mod.__dict__[name], type):
                     arg_type = PythonScraper.typename_to_typeref(mod_name, name)
                     break

--- a/Python/Product/Analyzer/PyLibAnalyzer.cs
+++ b/Python/Product/Analyzer/PyLibAnalyzer.cs
@@ -1363,7 +1363,11 @@ namespace Microsoft.PythonTools.Analysis {
                         _updater.UpdateStatus(_progressOffset, _progressTotal, "Saving " + currentLibrary);
                     }
                     Directory.CreateDirectory(outDir);
-                    new SaveAnalysis().Save(projectState, outDir);
+                    var saver = new SaveAnalysis();
+                    saver.Save(projectState, outDir);
+                    foreach (var err in saver.Errors) {
+                        TraceError("{0}", err);
+                    }
                     TraceInformation("End of group \"{0}\"", files[0].LibraryPath);
                     AnalysisLog.EndFileGroup();
 

--- a/Python/Product/Analyzer/PythonScraper.py
+++ b/Python/Product/Analyzer/PythonScraper.py
@@ -403,7 +403,7 @@ def generate_member(obj, is_hidden=False, from_type=False):
 
         # Callable objects with a docstring that provides us with at least one
         # overload will be treated as functions rather than data.
-        if hasattr(obj, '__call__'):
+        if safe_hasattr(obj, '__call__'):
             try:
                 info = generate_builtin_function(obj, from_type)
                 if info and info['overloads']:

